Compilation error in generated code for @Conditional and collection

When using mapstruct 1.5.2.Final and maven-compiler-plugin 3.8.1, we can produce a compilation error with the new @Conditional feature (which is very much appreciated in general).

Assume a JPA entity class CustomerPerson which has a Set<CustomerAddress>. Both should be mapped to CustomerPersonDTO containing a Set<CustomerAdressDTO>.

In our mapper, we have added code to prevent mapping addresses, when they are lazy from hibernate (see https://github.com/mapstruct/mapstruct/issues/2776)

```
@Mapper(uses = { CustomerAddressMapper.class }, config = MapStructConfiguration.class)
public interface CustomerPersonMapper {

	CustomerPersonDTO toDTO(final CustomerPerson entity);
	CustomerPerson toEntity(final CustomerPersonDTO dto);
	
	@Condition
    default boolean isLazyLoaded(Collection<?> sourceCollection) {

        // Case: Source field in domain object is lazy: Skip mapping
        if (!Hibernate.isInitialized(sourceCollection)) {
            // Skip mapping, as resetting the custom converter is not thread safe!
            return true;
        }

        // Continue mapping
        return false;
	
	}
}
```

Now the generated code in CustomerPersonMapperImpl contains a compilation error:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project mapstruct-bug: Compilation failure: Compilation failure:
[ERROR] /C:/gefa/git/gefa-incubator/mapstruct-bug/target/generated-sources/annotations/com/accenture/hpsapf/masterdatamanagement/mapper/CustomerPersonMapperImpl.java:[30,28] cannot find symbol
[ERROR]   symbol:   variable address
[ERROR]   location: class com.accenture.hpsapf.masterdatamanagement.mapper.CustomerPersonMapperImpl
[ERROR] /C:/gefa/git/gefa-incubator/mapstruct-bug/target/generated-sources/annotations/com/accenture/hpsapf/masterdatamanagement/mapper/CustomerPersonMapperImpl.java:[50,28] cannot find symbol
[ERROR]   symbol:   variable address
[ERROR]   location: class com.accenture.hpsapf.masterdatamanagement.mapper.CustomerPersonMapperImpl
[ERROR] -> [Help 1]
```

Reason is, that the lines are generated wrongly (or something with our usage of @Conditional is not correct):
```
...
 customerPersonDTO.setBirthday( entity.getBirthday() );
        if ( isLazyLoaded( address ) ) { /// <---- COMPILATION ERROR HERE, addresses is not known
            for ( CustomerAddress address : entity.getAddresses() ) {
                customerPersonDTO.addCustomerAddress( customerAddressMapper.toDTO( address ) );
            }
        }
```

In our opinion the generated code should look like this:

```
 customerPersonDTO.setBirthday( entity.getBirthday() );
        if ( isLazyLoaded( entity.getAddresses() ) ) {
            for ( CustomerAddress address : entity.getAddresses() ) {
                customerPersonDTO.addCustomerAddress( customerAddressMapper.toDTO( address ) );
            }
        }
```

You can find a self-contained simple maven test case to reproduce the issue attached: 
[mapstruct-bug.zip](https://github.com/mapstruct/mapstruct/files/9159579/mapstruct-bug.zip)

Is this a known issue? We could not find any similar reports.
Is there a workaround or a different usage for @Conditional? 

Thanks!
 Dominik

--- a/processor/src/main/resources/org/mapstruct/ap/internal/model/MethodReference.ftl
+++ b/processor/src/main/resources/org/mapstruct/ap/internal/model/MethodReference.ftl
@@ -64,6 +64,7 @@
 -->
 <#macro _assignment assignmentToUse>
    <@includeModel object=assignmentToUse
+        presenceCheck=ext.presenceCheck
         targetBeanName=ext.targetBeanName
                existingInstanceMapping=ext.existingInstanceMapping
                targetReadAccessorName=ext.targetReadAccessorName

--- a/processor/src/main/resources/org/mapstruct/ap/internal/model/MethodReferencePresenceCheck.ftl
+++ b/processor/src/main/resources/org/mapstruct/ap/internal/model/MethodReferencePresenceCheck.ftl
@@ -7,5 +7,6 @@
 -->
 <#-- @ftlvariable name="" type="org.mapstruct.ap.internal.model.MethodReferencePresenceCheck" -->
 <@includeModel object=methodReference
+               presenceCheck=true
                targetPropertyName=ext.targetPropertyName
                targetType=ext.targetType/>

--- a/processor/src/main/resources/org/mapstruct/ap/internal/model/common/SourceRHS.ftl
+++ b/processor/src/main/resources/org/mapstruct/ap/internal/model/common/SourceRHS.ftl
@@ -6,4 +6,4 @@
 
 -->
 <#-- @ftlvariable name="" type="org.mapstruct.ap.internal.model.common.SourceRHS" -->
-<#if sourceLoopVarName??>${sourceLoopVarName}<#elseif sourceLocalVarName??>${sourceLocalVarName}<#else>${sourceReference}</#if>
+<#if sourceLoopVarName?? && !ext.presenceCheck??>${sourceLoopVarName}<#elseif sourceLocalVarName??>${sourceLocalVarName}<#else>${sourceReference}</#if>

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2937/Issue2937Mapper.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2937/Issue2937Mapper.java
@@ -1,0 +1,62 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2937;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+import org.mapstruct.CollectionMappingStrategy;
+import org.mapstruct.Condition;
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+/**
+ * @author Filip Hrisafov
+ */
+@Mapper(collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED)
+public interface Issue2937Mapper {
+
+    Issue2937Mapper INSTANCE = Mappers.getMapper( Issue2937Mapper.class );
+
+    Target map(Source source);
+
+    @Condition
+    default boolean isApplicable(Collection<?> collection) {
+        return collection == null || collection.size() > 1;
+    }
+
+    class Source {
+        private final Collection<String> names;
+
+        public Source(Collection<String> names) {
+            this.names = names;
+        }
+
+        public Collection<String> getNames() {
+            return names;
+        }
+
+//        public boolean hasNames() {
+//            return names.size() > 1;
+//        }
+    }
+
+    class Target {
+        private final Collection<String> names;
+
+        public Target() {
+            this.names = new ArrayList<>();
+        }
+
+        public Collection<String> getNames() {
+            return names;
+        }
+
+        public void addName(String name) {
+            this.names.add( name );
+        }
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2937/Issue2937Mapper.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2937/Issue2937Mapper.java
@@ -39,9 +39,6 @@ public interface Issue2937Mapper {
             return names;
         }
 
-//        public boolean hasNames() {
-//            return names.size() > 1;
-//        }
     }
 
     class Target {

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2937/Issue2937Test.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2937/Issue2937Test.java
@@ -18,9 +18,9 @@ import static org.assertj.core.api.Assertions.assertThat;
  * @author Filip Hrisafov
  */
 @IssueKey("2937")
-@WithClasses( {
+@WithClasses({
     Issue2937Mapper.class,
-} )
+})
 class Issue2937Test {
 
     @ProcessorTest
@@ -37,6 +37,6 @@ class Issue2937Test {
         target = Issue2937Mapper.INSTANCE.map( source );
 
         assertThat( target.getNames() )
-            .containsExactly("Tester 1", "Tester 2");
+            .containsExactly( "Tester 1", "Tester 2" );
     }
 }

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2937/Issue2937Test.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2937/Issue2937Test.java
@@ -1,0 +1,42 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2937;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.mapstruct.ap.testutil.IssueKey;
+import org.mapstruct.ap.testutil.ProcessorTest;
+import org.mapstruct.ap.testutil.WithClasses;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Filip Hrisafov
+ */
+@IssueKey("2937")
+@WithClasses( {
+    Issue2937Mapper.class,
+} )
+class Issue2937Test {
+
+    @ProcessorTest
+    void shouldCorrectlyUseConditionalForAdder() {
+        List<String> sourceNames = new ArrayList<>();
+        sourceNames.add( "Tester 1" );
+        Issue2937Mapper.Source source = new Issue2937Mapper.Source( sourceNames );
+        Issue2937Mapper.Target target = Issue2937Mapper.INSTANCE.map( source );
+
+        assertThat( target.getNames() ).isEmpty();
+
+        sourceNames.add( "Tester 2" );
+
+        target = Issue2937Mapper.INSTANCE.map( source );
+
+        assertThat( target.getNames() )
+            .containsExactly("Tester 1", "Tester 2");
+    }
+}

